File: blue-core/src/main/java/blue/time/Meter.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2023 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */
package blue.time;

import electric.xml.Element;
import java.util.Objects;

/**
 * Represents a musical time signature (meter).
 * Immutable value object.
 * 
 * @author syi
 */
public class Meter {
    /** Minimum 1, max unbounded */
    public final long numBeats;
    
    /** Duration of beat, should be 1, 2, 4, 8, 16, 32 */
    public final long beatLength;  
    
    public Meter() {
        this(4, 4);
    }
    
    public Meter(long numBeats, long beatLength) {
        assert(numBeats > 0);
        assert(beatLength >= 0);
        
        this.numBeats = numBeats;
        this.beatLength = beatLength;
    }
    
    public Meter(Meter meter) {
        this.numBeats = meter.numBeats;
        this.beatLength = meter.beatLength;
    }
    
    /**
     * Calculates the duration of one measure in Csound beats (quarter notes).
     * For example, 4/4 time = 4 beats, 3/4 time = 3 beats, 6/8 time = 3 beats.
     * 
     * @return duration of one measure in Csound beats
     */
    public double getMeasureBeatDuration() {
        return numBeats * (4.0 / beatLength);
    }
    
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Meter)) return false;
        Meter other = (Meter) obj;
        return numBeats == other.numBeats && beatLength == other.beatLength;
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(numBeats, beatLength);
    }
    
    @Override
    public String toString() {
        return numBeats + "/" + beatLength;
    }
    
    /**
     * Save Meter to XML.
     */
    public Element saveAsXML() {
        Element retVal = new Element("meter");
        retVal.addElement("numBeats").setText(Long.toString(numBeats));
        retVal.addElement("beatLength").setText(Long.toString(beatLength));
        return retVal;
    }
    
    /**
     * Load Meter from XML.
     */
    public static Meter loadFromXML(Element data) {
        long numBeats = Long.parseLong(data.getElement("numBeats").getTextString());
        long beatLength = Long.parseLong(data.getElement("beatLength").getTextString());
        return new Meter(numBeats, beatLength);
    }
}

File: packages/blue-data/src/time/meter.ts
/**
 * Meter — an immutable time signature value object.
 * Mirrors the Java Meter class.
 *
 * For example, 4/4 time has numBeats=4, beatLength=4.
 * 6/8 time has numBeats=6, beatLength=8.
 */
import { Element } from '../serialization/xml-reader';

export class Meter {
  readonly numBeats: number;
  readonly beatLength: number;

  constructor(numBeats: number = 4, beatLength: number = 4) {
    this.numBeats = numBeats;
    this.beatLength = beatLength;
  }

  /**
   * Get the duration of one measure in Csound beats (quarter notes).
   * For 4/4: 4 * (4/4) = 4 beats.
   * For 3/4: 3 * (4/4) = 3 beats.
   * For 6/8: 6 * (4/8) = 3 beats.
   */
  getBeatsPerMeasure(): number {
    return this.numBeats * (4.0 / this.beatLength);
  }

  /**
   * Get the beat scale factor — how many Csound beats per one beat of this meter.
   * For /4: 4/4 = 1.0 (quarter note = 1 beat).
   * For /8: 4/8 = 0.5 (eighth note = 0.5 beats).
   */
  getBeatScale(): number {
    return 4.0 / this.beatLength;
  }

  /** Alias for getBeatsPerMeasure — matches Java getMeasureBeatDuration(). */
  getMeasureBeatDuration(): number {
    return this.getBeatsPerMeasure();
  }

  toString(): string {
    return `${this.numBeats}/${this.beatLength}`;
  }

  hashCode(): number {
    return (((this.numBeats * 31 + this.beatLength) >>> 0) * 2654435761) | 0;
  }

  equals(other: Meter): boolean {
    return this.numBeats === other.numBeats && this.beatLength === other.beatLength;
  }

  // ─── XML Serialization ───

  saveAsXML(): Element {
    const elem = new Element('meter');
    elem.addElement('numBeats').setText(this.numBeats.toString());
    elem.addElement('beatLength').setText(this.beatLength.toString());
    return elem;
  }

  static loadFromXML(data: Element): Meter {
    const numBeats = parseInt(data.getTextString('numBeats') ?? '4', 10);
    const beatLength = parseInt(data.getTextString('beatLength') ?? '4', 10);
    return new Meter(numBeats || 4, beatLength || 4);
  }
}
